tools/generate_torch_version.py

```diff
@@ -48,19 +49,29 @@ def get_tag(pytorch_root: str | Path) -> str:
 
 
 def get_torch_version(sha: str | None = None) -> str:
+    # If building from a source distribution, the PKG-INFO file will be present
+    # and will contain the version information.
+    # Otherwise, the version will be in version.txt, possibly extended with
+    # the git commit SHA.
     pytorch_root = Path(__file__).absolute().parent.parent
-    version = open(pytorch_root / "version.txt").read().strip()
-
-    if os.getenv("PYTORCH_BUILD_VERSION"):
-        assert os.getenv("PYTORCH_BUILD_NUMBER") is not None
-        build_number = int(os.getenv("PYTORCH_BUILD_NUMBER", ""))
-        version = os.getenv("PYTORCH_BUILD_VERSION", "")
-        if build_number > 1:
-            version += ".post" + str(build_number)
-    elif sha != UNKNOWN:
-        if sha is None:
-            sha = get_sha(pytorch_root)
-        version += "+git" + sha[:7]
+    pkg_info_path = pytorch_root / "PKG-INFO"
+    if pkg_info_path.exists():
+        with open(pkg_info_path) as f:
+            pkg_info = email.message_from_file(f)
+        version = pkg_info["Version"]
```

Comment:
I think the change can be as simple as https://github.com/pytorch/pytorch/pull/157432/files#diff-b276b28d4f6ce7bb5745d853b4791e1c9e2022801ade42a507fc6ac3d81bd874R60

The build info (string after the `+` sign in the full version) can contain various details, such as the Git SHA, CUDA build, etc.